File: Flink/src/main/scala/flink/study/batch/BatchWordCount.scala
package flink.study.batch

import org.apache.flink.api.scala.{DataSet, ExecutionEnvironment}
// 解决 flink wordcount error: No implicits found for parameter evidence$11: TypeInformation[S...
import org.apache.flink.streaming.api.scala._

/**
 * :Description:Flink batch demo.
 * :Author: 佳境Shmily
 * :Create Time: 2021/3/6 17:32
 * :Site: shmily-qjj.top
 */
object BatchWordCount {
  case class UserClickLog(userName:String, eventTime:String, logType:String, pageNum:Int)

  def main(args: Array[String]): Unit = {
    // 获取 批处理 执行环境  (批处理获取执行环境用ExecutionEnvironment，流处理获取环境用StreamExecutionEnvironment)
    val env:ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    //批处理后的数据是DataSet，流处理后的数据是DataStream.
    val logs:DataSet[UserClickLog] = env
      .fromElements(
        UserClickLog("user_1", "1000", "click", 1),
        UserClickLog("user_2", "2000", "click", 2),
        UserClickLog("user_1", "1500", "click", 2)
      )
    logs.map(x=>(x.userName,1)).groupBy(0).sum(1).print()

    // 在批处理中，如果输出目的端，执行的是print命令（除此之外，还有count，collect方法），则执行任务Execute不需要调用
    // 输出目的端调用writeAsCsv、writeAsText等其他方法，则后面需要调用Execute；
//    env.execute("BatchWordCount")

  }
}

File: Flink/src/main/java/flink/study/BatchWordCount.java
package flink.study;

import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.AggregateOperator;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.api.java.operators.FlatMapOperator;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.util.Collector;
import scala.Int;

import java.util.Arrays;

/**
 * :Description:Flink batchMode wordCount
 * :Author: 佳境Shmily
 * :Create Time: 2021/3/6 15:32
 * :Site: shmily-qjj.top
 */
public class BatchWordCount {
    public static void main(String[] args) throws Exception {
        //1.创建Env
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        //2.读文件 得到数据源
        DataSource<String> fileDataSource = env.readTextFile("D:\\Projects\\MyProjects\\IdeaProjects\\Shmily\\Flink\\input_data\\words.txt");

        //3.逻辑 数据扁平化打散flatMap得到单词word，word转为二元组(Tuple2)，统计元组
        // line->word->Tuple(word,1)->sum
        FlatMapOperator<String, Tuple2<String, Long>> wordTuple = fileDataSource.flatMap((String line, Collector<Tuple2<String, Long>> out) -> {
            String[] words = line.split(" ");
            Arrays.stream(words).forEach(word -> out.collect(Tuple2.of(word, 1L)));
        }).returns(Types.TUPLE(Types.STRING, Types.LONG));// Java原生不支持函数式编程，为了防止泛型擦除，需要加returns方法声明返回类型

        AggregateOperator<Tuple2<String, Long>> sum = wordTuple.groupBy(0).sum(1);

        //4 打印
        sum.print();

        env.execute();
    }
}
